Print algorithms

paskino opened the pull request:
This PR makes a couple of changes to the verbose output of an `Algorithm`.

 - `update_objective_interval` is used to control how often the objective function is evaluated. The iteration number and the objective are stored in the algorithm
 - the `run` method accepts the new `print_interval` parameter (defaults to `update_objective_interval`) which controls how often there is a print to screen during the run of the algorithm.
 - the `verbose` parameter is now an int: 0 means no print to screen (it was `verbose=False, very_verbose=False`), 1 medium verbosity and default setting (`verbose=True, very_verbose=False`), 2 highly verbose (`verbose=True, very_verbose=True`). The change is backward compatible and both `verbose` and `very_verbose` as boolean are accepted, though deprecated. 
 - `verbose=2` is a setting that is useful only for PDHG and SPDHG. However, the `verbose=2` setting is available for all algorithms, but in the column for the dual and primal-dual gap there will be printed `N/A`
 - the `callback` is evaluated at each `update_objective_interval` regardless of the `print_interval`
 - the timing of the time per iteration will be up-to-date, even if the objective isn't calculated
 - adds the `iterations` property to `Algorithm` where the iterations at which the objective has been evaluated are stored


Notice that print to screen will write the evaluated objective value only if the print happens at a multiple of `update_objective_interval`. Where not available it'll print `N/A`.

For instance

```python

algo = PrintAlgo(update_objective_interval = 10, max_iteration = 1000)

algo.run(20, verbose=2, print_interval = 2)

algo.run(3, verbose=1, print_interval = 2)

algo.run(20, verbose = 1, print_interval = 7)

algo.run(20, verbose=True, very_verbose=False)
algo.run(20, verbose=True, very_verbose=True, print_interval=7, callback=callback)

print (algo._iteration)
print (algo.objective)
```

This code outputs

```
     Iter   Max Iter     Time/Iter        Primal          Dual     Primal-Dual
                               [s]     Objective     Objective             Gap
        0       1000         0.000    0.00000e+00           N/A            N/A
        2       1000         0.100            N/A           N/A            N/A
        4       1000         0.100            N/A           N/A            N/A
        6       1000         0.100            N/A           N/A            N/A
        8       1000         0.100            N/A           N/A            N/A
       10       1000         0.100    1.00000e+02           N/A            N/A
       12       1000         0.100            N/A           N/A            N/A
       14       1000         0.100            N/A           N/A            N/A
       16       1000         0.100            N/A           N/A            N/A
       18       1000         0.100            N/A           N/A            N/A
----------------------------------------------------------------------------
       20       1000         0.100    1.00000e+02           N/A            N/A
Stop criterion has been reached.

     Iter   Max Iter     Time/Iter            Objective
                               [s]
       20       1000         0.100          4.00000e+02
       22       1000         0.100                  N/A
-------------------------------------------------------
       23       1000         0.100                  N/A
Stop criterion has been reached.

     Iter   Max Iter     Time/Iter            Objective
                               [s]
       23       1000         0.100                  N/A
       30       1000         0.100          9.00000e+02
       37       1000         0.100                  N/A
-------------------------------------------------------
       43       1000         0.100                  N/A
Stop criterion has been reached.

     Iter   Max Iter     Time/Iter            Objective
                               [s]
       43       1000         0.100                  N/A
       53       1000         0.100                  N/A
-------------------------------------------------------
       63       1000         0.100                  N/A
Stop criterion has been reached.

     Iter   Max Iter     Time/Iter        Primal          Dual     Primal-Dual
                               [s]     Objective     Objective             Gap
       63       1000         0.100            N/A           N/A            N/A
I am being called  70
       70       1000         0.100    4.90000e+03           N/A            N/A
       77       1000         0.100            N/A           N/A            N/A
I am being called  80
----------------------------------------------------------------------------
       83       1000         0.100            N/A           N/A            N/A
Stop criterion has been reached.

[0, 10, 20, 30, 40, 50, 60, 70, 80]
[0, 100, 400, 900, 1600, 2500, 3600, 4900, 6400]
```

paskino commented:
Also, notice that only `GradientDescent` requires the evaluation of the objective value at the initialisation, therefore the call to `update_objective` in `set_up` has been removed in all algorithms (this speeds up the set up of the algorithm).

paskino commented:
closes #669 
closes #673

jakobsj commented:
Looking at it now, it is looking very promising. Two immediate comments:

1. I would prefer to have blank space instead of N/A, that makes it simpler to look at.

2. This I think was present before this PR as I noticed the same yesterday. There seems to be a bug when stopping and starting the algorithm, specifically look at the output at iteration 20, when it stops, you have primal objective 1e+2. Then when it restarts it goes again iteration 20, but prints objective 4e+2. I don't mean the fact that it is bigger (which is also concerning, but I don't know what problem is being solved) but the fact that it is different. Has another iteration already been done, so it should print 21 in the first line instead of 20 again?

jakobsj commented:
There are some print outs that appear like debugging statements "I am being printed  70" and "I am being printed  80" in the last algorithm run. Remove?

paskino commented:
The lines `I am being called  80` are to demonstrate the use of the `callback` in the `run`. The `callback` function gets called every `update_objective_interval` (in the example `update_objective_interval = 10`, so it got called only at 70 and 80 iteration. In this case it just prints to screen.

Your point 2 before is well spotted. In this specific example the objective is the square of the iteration number. So, the 400 is correct and the 100 is wrong. This means that if you run 20 iterations it'll go from iteration 0 to 19. The actual objective printed is the last one to be available, i.e. 100 calculated at iteration 10.

Thanks I'll debug tomorrow and will remove `N/A`.

jakobsj commented:
OK great, thanks. A few follow-up points/questions:

What does iteration 0 mean?  Is that the first actual iteration? In other words what is stored in `alg.objective[0]` and `alg.objective[1]`?  I thought in position [0] we save the objective at the initial value, then run 20 iterations, so that alg.objective is 21 elements long, and the objective after iteration 20 is in position [20].

It's great to have the iterations property - that makes it easy to plot objective history vs the correct iteration numbers afterwards.

How would one specify NOT to save any objective values?  update_objective_interval = 0?

jakobsj commented:
Integers for verbose is great too: much easier to remember, and cleaner to add more verbosity levels if necessary later.

jakobsj commented:
With the timing being printed no matter whether the objective has been evaluated at that iteration, it should be possible to compare and see how much more expensive it is to compute the objective.

paskino commented:
> How would one specify NOT to save any objective values? update_objective_interval = 0?

Exactly

> What does iteration 0 mean? Is that the first actual iteration? In other words what is stored in alg.objective[0] and alg.objective[1]? I thought in position [0] we save the objective at the initial value, then run 20 iterations, so that alg.objective is 21 elements long, and the objective after iteration 20 is in position [20].

I think I've addressed your points. Now when you issue something like (notice that in this algorithm the objective function value is the square of the iteration number).
```python
algo = SomeAlgorithm(update_objective_interval=5, , max_iteration = 1000)
algo.run(10, print_interval=2, verbose=1)
print (algo.iterations)
print (algo.objective)

algo.run(10, print_interval=2, verbose=1)
print (algo.iterations)
print (algo.objective)

```
The expected behaviour is:
```
     Iter   Max Iter     Time/Iter        Primal          Dual     Primal-Dual
                               [s]     Objective     Objective             Gap
        0       1000         0.000    0.00000e+00
        2       1000         0.050
        4       1000         0.050
        6       1000         0.050
        8       1000         0.050
       10       1000         0.050    1.00000e+02
----------------------------------------------------------------------------
       10       1000         0.050    1.00000e+02
Stop criterion has been reached.

[0, 5, 10]
[0, 25, 100]
     Iter   Max Iter     Time/Iter            Objective
                               [s]
       11       1000         0.050
       13       1000         0.050
       15       1000         0.050          2.25000e+02
       17       1000         0.050
       19       1000         0.050
-------------------------------------------------------
       20       1000         0.050          4.00000e+02
Stop criterion has been reached.

[0, 5, 10, 15, 20]
[0, 25, 100, 225, 400]
```

I see:
 - in `algo.iterations` you find all the iterations at which the objective function has been evaluated, that's `[0,5,10,15,20]` as the `update_objective_interval = 5`
 - in `algo.objective` you find the relative values of the objective function
 - the first `run(10)` runs 10 iteration + the evaluation of the objective function on the starting solution, therefore the print stops at iteration 10.
 - the second `run(10)` starts from iteration 11 and proceeds with a `print_interval=2` apparently off sync. This is expected as the `print_interval` is detached from the `update_objective_interval` and is related to the issuing of `run`.

paskino commented:
@jakobsj Jenkins is happy, if you are too I will merge

jakobsj commented:
Looks great, thanks, please go ahead.

On Tue, 20 Oct 2020, 17:17 Edoardo Pasca, <notifications@github.com> wrote:

> @jakobsj <https://github.com/jakobsj> Jenkins is happy, if you are too I
> will merge
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/vais-ral/CCPi-Framework/pull/674#issuecomment-712928209>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACMDSCHLUGQ37RBO6PNIMODSLWSYLANCNFSM4SWNSXVQ>
> .
>
